mfsmanpages/mfschunktool.8
.TH mfschunktool "8" "May 2026" "MooseFS 4.59.2-1" "This is part of MooseFS"
.ss 12 0
.SH NAME
mfschunktool - checks chunk integrity offline
.SH SYNOPSIS
\fBmfschunktool\fP 
[\fB\-f\fP]
[\fB\-r\fP]
[\fB\-n\fP]
[\fB\-e\fP]
[\fB\-x\fP]
[\fB\-m\fP \fIdamaged_dir\fP]
\fIchunk_file\fP|\fImfs_hdd_path\fP ...
.SH DESCRIPTION
\fBmfschunktool\fP checks chunk integrity using stored CRC checksums in chunk's header.
It can also 'repair' chunk (fix stored checksums). This tool may be useful to check last
written chunks after possible data corruption (power down etc.)
If the supplied parameter is a file, the tool will check this file. If the supplied parameter is a directory, the tool will traverse this directory recursively and try to check all files that seem to be chunks (they are named like chunk files or, if \fB\-n\fP option is used, also have headers like chunk files, even when the name is wrong).
.PP
\fBNotice! This tool should never be used while the chunk server process is active!\fP If the tool is used on a chunk disk (aka path in mfshdd.cfg file) it will check if the chunk server is working and refuse to operate. But if it's used on a single file or subdirectory, such check is not possible and it's the responsibility of the user to make sure the tool is not used on "working" chunks, as it could lead to data loss.
.SH OPTIONS
.TP
\fB\-f\fP
turn on fast mode (in this mode only last block is tested - usually this is enough to find corrupted chunks)
.TP
\fB\-r\fP
turn on repair mode (if checksums on given chunk don't match then they will be overwritten by checksums calculated using actual data)
.TP
\fB\-n\fP
turn on name fixing mode (if the chunk file name is not correct but there is proper file header, the tool will rename given file using data stored in the header)
.TP
\fB\-e\fP
turn on empty block crc check for 1.0 version chunks (force checking crc values for non existing blocks - may fail in case of chunks that were truncated)
.TP
\fB\-x\fP
turn on reporting successful check (print 'OK' message when everything is correct)
.TP
\fB\-m\fP \fIdamaged_dir\fP
move damaged chunks to a specified directory \fIdamaged_dir\fP; see NOTES
.SH NOTES
.PP
The option to move damaged chunks to a specified directory is most useful when whole directories (disks) with chunks are scanned, and can be employed in different ways, depending on what other options are used at the same time.
.PP
If \fB\-r\fP option is used together with \fB\-m\fP option, then only chunks that are unrepairable will be moved. This is a fast approach to chunk repairing, that repairs and cleans the chunk directory (disk) in one go.
.PP
If only option \fB\-m\fP is used, then all damaged chunks will be moved to a specified directory. The user can then use the \fBmfschunktool\fP again in this directory, this time with the \fB\-r\fP option, to repair all chunks that can be repaired, but only introduce the repaired chunks back into the system if there is no other copy available from another chunkserver (a copy that did not need a repair operation in the first place). This is a more time consuming, but safer approach to repairing multiple chunks, e.g. after a power failure.
.SH "REPORTING BUGS"
Report bugs to <bugs@moosefs.com>.
.SH COPYRIGHT
Copyright (C) 2026 Jakub Kruszona-Zawadzki, Saglabs SA

This file is part of MooseFS.

MooseFS is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, version 2 (only).

MooseFS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, see
<https://www.gnu.org/licenses/>.
.SH "SEE ALSO"
.BR mfschunkserver (8)


NEWS
This file lists noteworthy changes in MooseFS.

* MooseFS 4.59.2-1 (2026-05-14)

  - (cs) fixed bug preventing background chunk testing

* MooseFS 4.59.1-1 (2026-05-04)

  - (cs) fixed potential OOB read

* MooseFS 4.59.0-1 (2026-04-23)

  - (all) changed licence handling (pro only)
  - (all) fixed some security issues

* MooseFS 4.58.5-1 (2026-04-01)

  - (cs) removed aggresive fadvise calls

* MooseFS 4.58.4-1 (2026-03-18)

  - (all) added spec files for Fedora, changed autogen.sh to bootstrap.sh
  - (all) added 'printflike' macro to mfslog, fixed all printflike macros, fixed some syslog args
  - (common) silenced address sanitizer unnecessary warning
  - (common) fixed masterresolver (potential segfault when the last name was empty - pro only)
  - (common) added protection against double initialization of strerr
  - (master) added expected chunk versions to syslog messages

* MooseFS 4.58.3-1 (2025-12-15)

  - (master+tools) fixed conditions for reversible storage class
  - (daemons) fixed small security issue (clearing auxiliary groups before setuid)
  - (libmfsio) fixed descriptor resize function

* MooseFS 4.58.2-1 (2025-11-03)

  - (master) changed timeouts used in master to master connections (pro only)
  - (client) fixed test for overlapping requests in read-ahead algorithm (pull request by Chuandev)
  - (all) fixed typos and spelling errors

* MooseFS 4.58.1-1 (2025-09-29)

  - (master) fix for segfault in flocklocks.c
  - (cli) fixed handling disconnected chunkservers
  - (cli) import fixes in cli mode
  - (tools) improved error messages in mfssetfacl
  - (tools) fixed parsing numerical groups in mfssetfacl
  - (master) prevent setting archive flag for files that are already in trash

* MooseFS 4.58.0-1 (2025-09-15)

  - (master) fixed support for partial readdir
  - (client) added reading dirs in parts
  - (client) added listxattr cache
  - (client) redesigned readdir caches
  - (client) added 'mfsreaddirplusminto' option needed for readdirplus

* MooseFS 4.57.8-1 (2025-08-05)

  - (crc) fixed crc calc for CPU's that can't access memory on non even addresses (i.e. arm-hf)
  - (cs) less disk ops for idle servers
  - (debian) updated postinstall scripts
  - (nbd) fixed restoring nbd devices after sleep/hibernate
  - (cli) fixed displaying master's and chunkserver's charts data
  - (systemd) added start/stop timeouts to chunkserver systemd parameters

* MooseFS 4.57.7-1 (2025-06-16)

  - (all) joined packages cgi and cgiserv to a new gui package
  - (gui) added dedicated GUI server
  - (gui) gui refactor
  - (cli) cli refactor
  - (gui) added support for Prometheus metrics
  - (master) fixed freeing waiting locks from data structures
  - (daemons) added condition to skip socket routines when poll returned zero
  - (daemons) fixed disabling OOM killer
  - (client) removed using potentially dangerous sprintf function

* MooseFS 4.57.6-1 (2025-03-25)

  - (master) fixed predefined storage classes
  - (master) fixed csipmap initialization issue (no csmap after start/restart)
  - (master) fixed lock listing (case of posix_locks and flocks being in the same packet)
  - (master) fixed chunk recovery conditions during read and lookup
  - (master) changed description of default storage classes
  - (man) added manpage for mfsmount.cfg file
  - (man) added subsections to all man pages and reworded some of them

* MooseFS 4.57.5-1 (2025-02-26)

  - (master) added the number of localsplit operations to the total split counter
  - (master) fixed chunk copies and EC matrix counters when both copies and parts are present
  - (master) fixed sending changelog rotate packet to metalogger (CE only)

* MooseFS 4.57.4-1 (2025-02-17)

  - (chunktool) added recursive mode and option to throw away damaged chunks

* MooseFS 4.57.3-1 (2025-02-04)

  - (mfsio) added support for xattr and acl
  - (master) added protection against setting too long xattr
  - (master) fixed case when user changes only grace period in quota

* MooseFS 4.57.2-1 (2025-01-27)

  - (mount) added 'fake' system.posix_acl* xattrs to listxattr (previously not needed)
  - (gui) added cluster traffic charts, updated names
  - (cli) added checks for wrong integer values in options
  - (freebsd) changed port base name in freebsd from moosefs4 to moosefs
  - (master) added KEEP_LEADERSHIP option
  - (gui) charts legend update

* MooseFS 4.57.1-1 (2025-01-08)

  - (master) fixed bug in case of making space for ec parts marked for removal
  - (gui) fixed tmp maintenance time display
  - (gui) fixed graph summary bar display
  - (master) fixed changelog rotation issues
  - (mfsio) fixed mutex initialisation issue
  - (autoconf) updated configure.ac for new PKG_PROG_PKG_CONFIG macro
  - (daemons) added timestamp to info file
  - (master) added meta info to info file
  - (gui) fixed asyncio for python <3.7

* MooseFS 4.57.0-1 (2024-12-09)

  - (all) added storage class priority, description and export group
  - (master) added storage class group allowance to exports.cfg (deprecates mingoal/maxgoal)
  - (master) changed predefined storage classes in new MFS instances
  - (all) removed support for python 2.7
  - (master) always resolve master hostname on info command (PRO only)
  - (master) added to supervisor in 'simple mode' broadcasting LEADER IP to all FOLLOWERs (PRO only)
  - (master) added info for changelog.c
  - (daemons) warnings if a variable is defined more than once in a config file
  - (bdev) added test for posix compliant ioctl prototype
  - (freebsd) moved port files from sysutils to filesystems
  - (master) fixed chunkserver registration in DEPUTY mode (PRO only)
  - (all) added names to some constants used across the system
  - (all) added new charts data (total data I/O from all clients)
  - (metasearch) added printing paths for trash and sustained items

mfschunkserver/mfschunktool.c
/*
 * Copyright (C) 2026 Jakub Kruszona-Zawadzki, Saglabs SA
 * 
 * This file is part of MooseFS.
 * 
 * MooseFS is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 (only).
 * 
 * MooseFS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see
 * <https://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <dirent.h>
#include <limits.h>
#include <errno.h>

#include "crc.h"
#include "datapack.h"
#include "clocks.h"
#include "MFSCommunication.h"
#include "idstr.h"

#define CHUNKCRCSIZE 4096

#define MODE_FAST 1
#define MODE_EMPTY 2
#define MODE_NAME 4
#define MODE_REPAIR 8

#define NAME_NONE 0
#define NAME_MFSDIR 1
#define NAME_CHUNK 2
#define NAME_ANY 3
#define NAME_ANYDIR 4
#define NAME_FIRST 5

static uint64_t lasttime = 0;
static uint64_t scanned = 0;
static uint32_t stdnlneeded = 0;
static uint32_t errnlneeded = 0;

static inline int check_timer(void) {
	uint64_t nsec;
	nsec = monotonic_nseconds();
	if (nsec>lasttime+1000000000 || nsec<lasttime) {
		lasttime = nsec;
		return 1;
	}
	return 0;
}

static inline void check_nl(void) {
	if (stdnlneeded) {
		printf("\33[2K\r"); // try to erase whole line using ANSI code
		while (stdnlneeded) {
			printf(" ");
			stdnlneeded--;
		}
		printf("\r");
	}
	if (errnlneeded) {
		fprintf(stderr,"\33[2K\r"); // try to erase whole line using ANSI code
		while (errnlneeded) {
			fprintf(stderr," ");
			errnlneeded--;
		}
		fprintf(stderr,"\r");
	}
}

static inline void move_file(const char *oldname,const char *newname,uint8_t *buff) {
	int ofd;
	int nfd;
	off_t s,pos;
	ssize_t bsize;

	ofd = open(oldname,O_RDONLY);
	if (ofd<0) {
		check_nl();
		printf("%s: error opening file !!!\n",oldname);
		return;
	}
	nfd = open(newname,O_RDWR|O_CREAT|O_EXCL,0666);
	if (nfd<0) {
		check_nl();
		printf("%s: error creating file !!!\n",newname);
		close(ofd);
		return;
	}

	s = lseek(ofd,0,SEEK_END);
	lseek(ofd,0,SEEK_SET);
	pos = 0;
	while (pos<s) {
		bsize = s - pos;
		if (bsize>MFSBLOCKSIZE) {
			bsize = MFSBLOCKSIZE;
		}
		if (read(ofd,buff,bsize)!=bsize) {
			check_nl();
			printf("%s: error reading file !!!\n",oldname);
			close(ofd);
			close(nfd);
			unlink(newname);
			return;
		}
		if (write(nfd,buff,bsize)!=bsize) {
			check_nl();
			printf("%s: error writing file !!!\n",newname);
			close(ofd);
			close(nfd);
			unlink(newname);
			return;
		}
		pos += bsize;
	}
	close(ofd); // we don't need to check close status here since the file has been correctly read
	if (close(nfd)<0) {
		check_nl();
		printf("%s: error writing file !!!\n",newname);
		close(nfd);
		unlink(newname);
		return;
	}
	if (unlink(oldname)<0) {
		printf("%s: error removing file !!!\n",oldname);
	}
}

static inline int hdd_check_filename(const char *fname,uint64_t *chunkid,uint32_t *version) {
	uint64_t namechunkid;
	uint32_t nameversion;
	char ch;
	uint32_t i;

	if (strncmp(fname,"chunk_",6)!=0) {
		return -1;
	}
	namechunkid = 0;
	nameversion = 0;
	for (i=6 ; i<22 ; i++) {
		ch = fname[i];
		if (ch>='0' && ch<='9') {
			ch-='0';
		} else if (ch>='A' && ch<='F') {
			ch-='A'-10;
		} else {
			return -1;
		}
		namechunkid *= 16;
		namechunkid += ch;
	}
	if (fname[22]!='_') {
		return -1;
	}
	for (i=23 ; i<31 ; i++) {
		ch = fname[i];
		if (ch>='0' && ch<='9') {
			ch-='0';
		} else if (ch>='A' && ch<='F') {
			ch-='A'-10;
		} else {
			return -1;
		}
		nameversion *= 16;
		nameversion += ch;
	}
	if (strcmp(fname+31,".mfs")!=0) {
		return -1;
	}
	if (chunkid!=NULL) {
		*chunkid = namechunkid;
	}
	if (version!=NULL) {
		*version = nameversion;
	}
	return 0;
}

int chunk_repair(const char *fname,char **newname,uint8_t mode,uint8_t showok,uint8_t *buff) {
	uint64_t namechunkid;
	uint32_t nameversion;
	uint32_t i,j;
	int fd;
	uint32_t crc[1024];
	uint32_t crcblock;
	uint16_t hdrsize;
	off_t s;
	const uint8_t *rp;
	uint8_t *wp;
	int ret=0;

	*newname = NULL;

	// check fname
	// name should be in format: ..../chunk_XXXXXXXXXXXXXXXX_YYYYYYYY.mfs
	if (mode&MODE_REPAIR) {
		fd = open(fname,O_RDWR);
	} else {
		fd = open(fname,O_RDONLY);
	}
	if (fd<0) {
		check_nl();
		printf("%s: error opening file !!!\n",fname);
		return ret | 16;
	}
	i = strlen(fname);
	if (i<35 || hdd_check_filename(fname+(i-35),&namechunkid,&nameversion)<0) {
		if (mode&MODE_NAME) {
			check_nl();
			printf("%s: wrong chunk name - try to fix it using header data\n",fname);
			if (read(fd,buff,20)!=20) {
				printf("%s: error reading header !!!\n",fname);
				close(fd);
				return ret | 16;
			}
			if (memcmp(buff,MFSSIGNATURE "C 1.",7)!=0 || (buff[7]!='0' && buff[7]!='1')) {
				printf("%s: wrong chunk header !!!\n",fname);
				close(fd);
				return ret | 16;
			}
			if (buff[7]=='1') { // this is chunk 1.1 - check 'empty' crc values
				mode |= MODE_EMPTY;
			}
			rp = buff+8;
			namechunkid = get64bit(&rp);
			nameversion = get32bit(&rp);
			j = i;
			while (j>0 && fname[j-1]!='/') {j--;}
			*newname = malloc(j+35+1);
			if (j>0) {
				memcpy(*newname,fname,j);
			}
			snprintf((*newname)+j,35+1,"chunk_%016"PRIX64"_%08"PRIX32".mfs",namechunkid,nameversion);
			(*newname)[j+35] = 0;
			if (rename(fname,(*newname))<0) {
				printf("%s->%s: rename error !!!\n",fname,(*newname));
				free(*newname);
				*newname = NULL;
				close(fd);
				return ret | 16;
			}
			printf("%s: changed name to: %s\n",fname,*newname);
		} else {
			check_nl();
			printf("%s: wrong chunk name format !!! (skip header)\n",fname);
			ret |= 1;
		}
	} else {
		*newname = strdup(fname);
		if (read(fd,buff,20)!=20) {
			check_nl();
			printf("%s: error reading header !!!\n",fname);
			close(fd);
			return ret | 16;
		}
		if (memcmp(buff,MFSSIGNATURE "C 1.",7)!=0 || (buff[7]!='0' && buff[7]!='1')) {
			check_nl();
			printf("%s: wrong chunk header !!!\n",fname);
			memcpy(buff,MFSSIGNATURE "C 1.0",8);
			ret |= 1;
		}
		if (buff[7]=='1') { // this is chunk 1.1 - check 'empty' crc values
			mode |= MODE_EMPTY;
		}
		rp = buff+8;
		wp = (uint8_t*)rp;
		if (get64bit(&rp)!=namechunkid) {
			check_nl();
			printf("%s: wrong chunk number in header !!!\n",fname);
			put64bit(&wp,namechunkid);
			ret |= 1;
		}
		wp = (uint8_t*)rp;
		if (get32bit(&rp)!=nameversion) {
			check_nl();
			printf("%s: wrong chunk version in header !!!\n",fname);
			put32bit(&wp,nameversion);
			ret |= 1;
		}
		if ((mode&MODE_REPAIR) && (ret&1)) {
			if (lseek(fd,0,SEEK_SET)!=0) {
				check_nl();
				printf("%s: error setting file pointer\n",fname);
				close(fd);
				return ret | 16;
			}
			if (write(fd,buff,20)!=20) {
				check_nl();
				printf("%s: error writing header !!!\n",fname);
				close(fd);
				return ret | 16;
			}
			ret |= 4;
		}
	}

	s = lseek(fd,0,SEEK_END);
	hdrsize = (s - CHUNKCRCSIZE) & MFSBLOCKMASK;
	if (hdrsize!=1024 && hdrsize!=4096) {
		check_nl();
		printf("%s: wrong file size\n",fname);
		close(fd);
		return ret | 16;
	}
	// read crc
	if (lseek(fd,hdrsize,SEEK_SET)!=hdrsize) {
		check_nl();
		printf("%s: error setting file pointer\n",fname);
		close(fd);
		return ret | 16;
	}
	if (read(fd,buff,CHUNKCRCSIZE)!=CHUNKCRCSIZE) {
		check_nl();
		printf("%s: error reading checksum block\n",fname);
		close(fd);
		return ret | 16;
	}
	rp = buff;
	for (i=0 ; i<1024 ; i++) {
		crc[i] = get32bit(&rp);
	}

	// check data crc
	if ((mode&(MODE_FAST|MODE_REPAIR))==MODE_FAST) {
		if (s<MFSBLOCKSIZE) { // this is empty file
			s = -1;
		} else {
			s = lseek(fd,-MFSBLOCKSIZE,SEEK_END);
			if (s<(hdrsize + CHUNKCRCSIZE)) {
				check_nl();
				printf("%s: wrong file size\n",fname);
				close(fd);
				return ret | 16;
			}
			s -= (hdrsize + CHUNKCRCSIZE);
			if ((s%MFSBLOCKSIZE)!=0) {
				check_nl();
				printf("%s: wrong file size\n",fname);
				close(fd);
				return ret | 16;
			}
			s >>= 16;
			if (read(fd,buff,MFSBLOCKSIZE)!=MFSBLOCKSIZE) {
				check_nl();
				printf("%s: error reading last data block\n",fname);
				close(fd);
				return ret | 16;
			}
			crcblock = mycrc32(0,buff,MFSBLOCKSIZE);
			if (crc[s]!=crcblock) {
				check_nl();
				printf("%s: crc error (last block ; header crc: %08"PRIX32" ; block crc: %08"PRIX32")\n",fname,crc[s],crcblock);
				ret |= 2;
			}
		}
		if ((mode&MODE_EMPTY) && s<1023) {
			if (crc[s+1]!=mycrc32_zeroblock(0,MFSBLOCKSIZE) && crc[s+1]!=0) {
				check_nl();
				printf("%s: crc error (first empty block (%u) has 'non zero' crc: %08"PRIX32")\n",fname,(unsigned int)(s+1),crc[s+1]);
				ret |= 2;
			}
		}
	} else {
		if (lseek(fd,(hdrsize + CHUNKCRCSIZE),SEEK_SET)!=(hdrsize + CHUNKCRCSIZE)) {
			check_nl();
			printf("%s: error setting file pointer\n",fname);
			close(fd);
			return ret | 16;
		}
		for (i=0 ; i<1024 ; i++) {
			s = read(fd,buff,MFSBLOCKSIZE);
			if (s==0) {
				crcblock = mycrc32_zeroblock(0,MFSBLOCKSIZE);
				if ((mode&MODE_EMPTY)==0 || crc[i]==0) {
					crc[i] = crcblock;
				}
			} else {
				if (s!=MFSBLOCKSIZE) {
					check_nl();
					printf("%s: error reading data block: %"PRIu32"\n",fname,i);
					close(fd);
					return ret | 16;
				}
				crcblock = mycrc32(0,buff,MFSBLOCKSIZE);
			}
			if (crc[i]!=crcblock) {
				check_nl();
				printf("%s: crc error (block: %"PRIu32" ; header crc: %08"PRIX32" ; block crc: %08"PRIX32")\n",fname,i,crc[i],crcblock);
				crc[i] = crcblock;
				ret |= 2;
			}
		}
		if ((mode&MODE_REPAIR) && (ret&2)) {
			// write crc
			wp = buff;
			for (i=0 ; i<1024 ; i++) {
				put32bit(&wp,crc[i]);
			}
			if (lseek(fd,hdrsize,SEEK_SET)!=hdrsize) {
				check_nl();
				printf("%s: error setting file pointer\n",fname);
				close(fd);
				return ret | 16;
			}
			if (write(fd,buff,CHUNKCRCSIZE)!=CHUNKCRCSIZE) {
				check_nl();
				printf("%s: error writing checksum block\n",fname);
				close(fd);
				return ret | 16;
			}
			ret |= 8;
		}
	}
	close(fd);
	if (ret==0 && showok) {
		check_nl();
		printf("%s: OK\n",fname);
	}
	if (ret&4) {
		check_nl();
		printf("%s: header fixed\n",fname);
		if (*newname!=NULL) {
			free(*newname);
		}
		*newname = NULL; // do not move fixed files !!!
	}
	if (ret&8) {
		check_nl();
		printf("%s: crc fixed\n",fname);
		if (*newname!=NULL) {
			free(*newname);
		}
		*newname = NULL; // do not move fixed files !!!
	}
	return ret;
}

static inline int is_hex(char c) {
	return ((c>='0' && c<='9') || (c>='a' && c<='f') || (c>='A' && c<='F'));
}

int recursive_scan(const char *path,uint8_t pathtype,uint8_t mode,uint8_t verbose,uint8_t *auxbuff,const char *dmgpath) {
	struct stat st;
	int lfd;
	DIR *dd;
	struct dirent *de;
#if !defined(__GLIBC__) || (__GLIBC__ < 2) || ((__GLIBC__ == 2) && (__GLIBC_MINOR__ < 23))
	struct dirent *destorage;
#endif
	char *subpath;
	char *newname;
	uint32_t dleng;
	uint32_t pleng;
	uint32_t fnleng;
	uint8_t goodname;
	uint8_t ret = 0;

	pleng = strlen(path);
	if (verbose==0 && check_timer()) {
		if (isatty(fileno(stdout))) {
			check_nl();
			printf("\robjects scanned: %"PRIu64" (path: %s) ... ",scanned,path);
			stdnlneeded = pleng+40;
			fflush(stdout);
		} else if (isatty(fileno(stderr))) {
			check_nl();
			fprintf(stderr,"\robjects scanned: %"PRIu64" (path: %s) ... ",scanned,path);
			errnlneeded = pleng+40;
			fflush(stderr);
		}
	}
	if (dmgpath!=NULL) {
		dleng = strlen(dmgpath);
		if (dleng<=pleng && memcmp(path,dmgpath,dleng)==0) { // path is inside of dmgpath
			return 0;
		}
	} else {
		dleng = 0;
	}
	if (stat(path,&st)<0) {
		check_nl();
		printf("%s: can't stat path\n",path);
		return 1;
	} else {
		if (S_ISDIR(st.st_mode) && (pathtype==NAME_MFSDIR || pathtype==NAME_ANY || pathtype==NAME_ANYDIR || pathtype==NAME_FIRST)) {
			subpath = malloc(pleng+1+256);
			if (subpath==NULL) {
				check_nl();
				printf("out of memory\n");
				return 1;
			}
			memcpy(subpath,path,pleng);
			subpath[pleng]='/';
			memcpy(subpath+pleng+1,".lock",5);
			subpath[pleng+6]=0;
			lfd = open(subpath,O_RDWR|O_CREAT|O_TRUNC,0640);
			if (lfd>=0 && lockf(lfd,F_TLOCK,0)<0) {
				check_nl();
				printf("found active lock file: %s - there is a working chunkserver using this disk - skip scanning\n",subpath);
				close(lfd);
				free(subpath);
				return 1;
			}
			dd = opendir(path);
			if (dd==NULL) {
				check_nl();
				printf("%s: can't open directory\n",path);
				if (lfd>=0) {
					close(lfd);
				}
				free(subpath);
				return 1;
			}

			/* size of name added to size of structure because on some os'es d_name has size of 1 byte */
#if !defined(__GLIBC__) || (__GLIBC__ < 2) || ((__GLIBC__ == 2) && (__GLIBC_MINOR__ < 23))
			destorage = (struct dirent*)malloc(sizeof(struct dirent)+pathconf(path,_PC_NAME_MAX)+1);
			if (destorage==NULL) {
				check_nl();
				printf("out of memory\n");
				closedir(dd);
				if (lfd>=0) {
					close(lfd);
				}
				free(subpath);
				return 1;
			}
			while (readdir_r(dd,destorage,&de)==0 && de!=NULL) {
#else
			while ((de = readdir(dd)) != NULL) {
#endif
				fnleng = strlen(de->d_name);
				if (fnleng==1 && de->d_name[0]=='.') {
					continue;
				}
				if (fnleng==2 && de->d_name[0]=='.' && de->d_name[1]=='.') {
					continue;
				}
				if (fnleng==5 && memcmp(de->d_name,".lock",5)==0) {
					continue;
				}
				if (fnleng==7 && memcmp(de->d_name,".metaid",7)==0) {
					continue;
				}
				if (fnleng==8 && memcmp(de->d_name,".chunkdb",8)==0) { // TODO: in the future read this file and compare contents with actual chunk list - also fix this file !!!
					continue;
				}
				goodname = NAME_NONE;
				if ((mode&MODE_NAME)) {
					goodname = NAME_ANY;
				} else {
					if (fnleng==2 && is_hex(de->d_name[0]) && is_hex(de->d_name[1]) && (pathtype==NAME_ANYDIR || pathtype==NAME_FIRST)) {
						goodname = NAME_MFSDIR;
					} else if (fnleng==35 && hdd_check_filename(de->d_name,NULL,NULL)>=0 && (pathtype==NAME_MFSDIR || pathtype==NAME_FIRST)) {
						goodname = NAME_CHUNK;
					} else if (pathtype==NAME_ANYDIR || pathtype==NAME_FIRST) {
						goodname = NAME_ANYDIR;
					}
				}
				if (goodname != NAME_NONE) {
					if (fnleng>255) {
						check_nl();
						printf("%s/%s: name too long\n",path,de->d_name);
						ret |= 1;
					} else {
						memcpy(subpath+pleng+1,de->d_name,fnleng);
						subpath[pleng+fnleng+1]=0;
						ret |= recursive_scan(subpath,goodname,mode,verbose,auxbuff,dmgpath);
					}
				}
			}
#if !defined(__GLIBC__) || (__GLIBC__ < 2) || ((__GLIBC__ == 2) && (__GLIBC_MINOR__ < 23))
			free(destorage);
#endif
			closedir(dd);
			if (lfd>=0) {
				close(lfd);
			}
			free(subpath);
		} else if (S_ISREG(st.st_mode) && (pathtype==NAME_CHUNK || pathtype==NAME_ANY || pathtype==NAME_FIRST)) {
			ret = chunk_repair(path,&newname,mode,verbose,auxbuff);
			scanned++;
			if (ret!=0 && dmgpath!=NULL && newname!=NULL) { // rename newname -> dmgpath/chunk_XXXXXXXXXXXXXXXX_YYYYYYYY.mfs
				fnleng = strlen(newname);
				if (fnleng<35) {
					check_nl();
					printf("%s: wrong chunk file name\n",newname);
					free(newname);
					return ret;
				}
				subpath = malloc(dleng+35+1+1);
				if (subpath==NULL) {
					check_nl();
					printf("out of memory\n");
					free(newname);
					return ret;
				}
				memcpy(subpath,dmgpath,dleng);
				subpath[dleng]='/';
				memcpy(subpath+dleng+1,newname+(fnleng-35),35);
				subpath[dleng+35+1]=0;
				if (rename(newname,subpath)<0) {
					if (errno==EXDEV) {
						move_file(newname,subpath,auxbuff);
					} else {
						check_nl();
						printf("%s -> %s: rename error !!!\n",newname,subpath);
					}
				}
				free(subpath);
			}
			if (newname!=NULL) {
				free(newname);
			}
			return ret;
		} else {
			check_nl();
			if (mode&MODE_NAME) {
				printf("%s: not a directory nor a file\n",path);
			} else {
				printf("%s: not a correct moosefs path\n",path);
			}
			return 1;
		}
	}
	return ret;
}

void usage(const char *appname) {
	fprintf(stderr,"usage: %s [-frnex] [-m damaged_dir] chunk_file|mfs_hdd_path ...\n",appname);
	fprintf(stderr,"\n");
	fprintf(stderr,"-f: fast check (check only header and crc of last data block)\n");
	fprintf(stderr,"-r: repair (fix header info from file name and recalculate crc)\n");
	fprintf(stderr,"-n: when file name is wrong then try to fix it using header\n");
	fprintf(stderr,"-e: force checking crc values for non existing blocks in chunks 1.0 - may fail in case of chunks that were truncated\n");
	fprintf(stderr,"-x: print 'OK' for good files\n");
	fprintf(stderr,"-m damaged_dir: move all damaged chunks to given directory for future processing\n");
}

int main(int argc,char *argv[]) {
	int ch;
	uint8_t mode = 0;
	uint8_t verbose = 0;
	uint8_t ret = 0;
	uint8_t *auxbuff;
	const char *appname;
	char *dmgpath;
	char *wrkpath;
	struct stat st;

	dmgpath = NULL;
	appname = argv[0];
	while ((ch = getopt(argc,argv,"hvfenrm:x?")) != -1) {
		switch(ch) {
			case 'h':
				usage(appname);
				return 0;
			case 'v':
				printf("version: %s\n",VERSSTR);
				return 0;
			case 'f':
				mode |= MODE_FAST;
				break;
			case 'e':
				mode |= MODE_EMPTY;
				break;
			case 'n':
				mode |= MODE_NAME;
				break;
			case 'r':
				mode |= MODE_REPAIR;
				break;
			case 'm':
				if (dmgpath!=NULL) {
					free(dmgpath);
				}
				dmgpath = realpath(optarg,NULL);
				if (dmgpath==NULL) {
					fprintf(stderr,"%s: realpath error\n",optarg);
					return 1;
				}
				if (stat(dmgpath,&st)<0) {
					fprintf(stderr,"%s: stat error\n",dmgpath);
					return 1;
				}
				if (!S_ISDIR(st.st_mode)) {
					fprintf(stderr,"%s: not a directory\n",dmgpath);
					return 1;
				}
				break;
			case 'x':
				verbose++;
				break;
//			case 'l':
//				if (bindhost) {
//					free((char*)bindhost);
//				}
//				bindhost = strdup(optarg);
//				break;
			default:
				usage(argv[0]);
				return 1;
		}
	}
	argc -= optind;
	argv += optind;

	if (argc==0) {
		usage(appname);
		if (dmgpath!=NULL) {
			free(dmgpath);
		}
		return 0;
	}

	mycrc32_init();

	auxbuff = malloc(MFSBLOCKSIZE);

	check_timer();

	while (argc>0) {
		wrkpath = realpath(*argv,NULL);
		if (wrkpath==NULL) {
			check_nl();
			printf("%s: realpath error\n",*argv);
			ret |= 1;
		} else {
			ret |= recursive_scan(wrkpath,NAME_FIRST,mode,verbose,auxbuff,dmgpath);
			free(wrkpath);
		}
		argv++;
		argc--;
	}

	free(auxbuff);
	if (dmgpath!=NULL) {
		free(dmgpath);
	}

	check_nl();

	return ret;
}
